test/puppetlabs/dujour/version_check_test.clj

```diff
@@ -0,0 +1,61 @@
+(ns puppetlabs.dujour.version-check-test
+  (:require [clojure.test :refer :all]
+            [puppetlabs.trapperkeeper.testutils.webserver :as jetty9]
+            [puppetlabs.trapperkeeper.testutils.logging :refer [with-test-logging]]
+            [puppetlabs.dujour.version-check :refer :all]
+            [schema.test :as schema-test]
+            [cheshire.core :as json]))
+
+(use-fixtures :once schema-test/validate-schemas)
+
+(deftest test-get-coords
+  (testing "group-id should use the default if not specified"
+    (is (= {:group-id    "puppetlabs.packages"
+            :artifact-id "foo"}
+           (get-coords "foo"))))
+  (testing "should use group-id if specified"
+    (is (= {:group-id    "foo.foo"
+            :artifact-id "foo"}
+           (get-coords {:group-id "foo.foo"
+                        :artifact-id "foo"})))))
+
+(defn update-available-app
+  [req]
+  {:status 200
+   :body (json/generate-string {:newer true
+                                :link "http://foo.com"
+                                :message "Howdy!"
+                                :product "foo"
+                                :version "9000.0.0"})})
+
+(defn server-error-app
+  [req]
+  {:status 500
+   :body "aaaaaaaaaaaaaaaaaaaaaaaaaa"})
+
+(deftest test-check-for-updates!
+  (with-test-logging
+    (jetty9/with-test-webserver update-available-app port
+                                (check-for-updates! "foo" (format "http://localhost:%s" port))
+                                (is (logged? #"Newer version 9000.0.0 is available!" :info))))
+  (with-test-logging
+    (jetty9/with-test-webserver server-error-app port
+                                (check-for-updates! "foo" (format "http://localhost:%s" port))
+                                (is (logged? #"Could not retrieve update information" :debug)))))
+
+(deftest test-version-check
+  (with-test-logging
+    (jetty9/with-test-webserver update-available-app port
+                                (version-check "foo" (format "http://localhost:%s" port))
+                                (Thread/sleep 1000)
```

Comment:
I agree with @jeffmccune that this is sort of a code smell, and we should at least be considering options for how we might avoid doing a sleep here.  Especially one this lengthy.

First thought: since this is a tiny library and probably always will be, it probably won't be being modified very often and the tests won't be run very often.  So if we have to leave in some sleeps in the tests, it's not the end of the world... but still not awesome.  I'd be much more adamant about addressing this if it was directly in the Puppet Server codebase.

@jeffmccune 's suggestion of changing the code such that you'd inject a logger object is very reasonable and I'd probably vote for that, except there is no object or protocol that we use for logging.  They're just functions, and they live in a namespace that we don't own.  So to do that, we'd have to create our own logging protocol and then pass in something that satisfied it, and we'd have basically introduced a second logging API at that point.  That seems a bit heavy-handed for this use case, though the general idea is quite sound to keep in mind for other use cases in the future.

That said, I can think of three options:

* Just test the core functions and not the async functions.  Not awesome, but if the async function is simple enough to reason about by just looking at the code, might be OK.
* Use with-redefs to override the log functions during testing, and then do something like modify an atom or a promise in your alternate implementation, and test that.
* Use an agent instead of a future in the async implementation and then use a watch on the agent in your test.  This would probably be my preferred option.  The only downside of it would be that dragging the whole clojure agent system into a codebase just for a simple problem like this might also be a little heavy-handed.  However, I doubt that we will ever have an application that is sophisticated enough to want to check in w/dujour and not already be using agents.  We already interact with it a little in TK itself and probably will have a few more places where we use it in TK in the future, so any TK app probably already meets the threshold... so I think it'd be OK to use them here, and that's probably what I'd do.  (I probably should have done that in the first place in the code you copied from Puppet Server, but it was a rush job copying and pasting that stuff from PuppetDB at the last minute before we did our first OSS release.)